src/test/scala/org/constellation/domain/cloud/CloudServiceTest.scala

```diff
@@ -0,0 +1,101 @@
+
+  before {
+    metrics = mock[Metrics]
+    metrics.updateMetricAsync(*, any[Long]) shouldReturnF (())
+    cloudService = CloudService(CloudConfig.empty, metrics)
+    cloudServiceEnqueue = cloudService.cloudSendingQueue(Queue.unbounded[IO, DataToSend].unsafeRunSync()).unsafeRunSync()
+  }
+
+  "verifyAndSend" - {
+    "verify the previous sent snapshot correctly when there is no snapshot" in {
+      val result = cloudService.verifyAndSend(2L, "somehash", IO.delay(List(().asRight[Throwable]))).unsafeRunSync()
+
+      result shouldBe List(Right(()))
+    }
+
+    "verify the previous sent snapshot correctly when the chain is preserved" in {
+      val previousSnapshot = SnapshotSent(10L, "somehash", true, true).some
+      cloudService.lastSentSnapshot.set(previousSnapshot).unsafeRunSync()
+
+      val result = cloudService.verifyAndSend(12L, "otherHash", IO.delay(List(().asRight[Throwable]))).unsafeRunSync()
+      val expected = List(Right(()))
+
+      result shouldBe expected
+    }
+
+    "verify the previous sent snapshot correctly when we are sending data for the same snapshot" in {
+      val previousSnapshot = SnapshotSent(10L, "somehash", true, true).some
+      cloudService.lastSentSnapshot.set(previousSnapshot).unsafeRunSync()
+
+      val result = cloudService.verifyAndSend(10L, "somehash", IO.delay(List(().asRight[Throwable]))).unsafeRunSync()
+      val expected = List(Right(()))
+
+      result shouldBe expected
+    }
+
+    "fail verification if previously sent snapshot is not directly preceding the one we are sending" in {
+      val previousSnapshot = SnapshotSent(10L, "somehash", true, true).some
+      cloudService.lastSentSnapshot.set(previousSnapshot).unsafeRunSync()
+
+      val result = cloudService.verifyAndSend(14L, "otherHash", IO.delay(List(().asRight[Throwable]))).unsafeRunSync()
+      val expected = List(Left(PrecedingSnapshotIsNotThePreviousOne(previousSnapshot, 14L, "otherHash")))
+
+      result shouldBe expected
+    }
+
+    "fail verification if previously sent snapshot is on the same height but the hash is different" in {
+      val previousSnapshot = SnapshotSent(10L, "somehash", true, true).some
+      cloudService.lastSentSnapshot.set(previousSnapshot).unsafeRunSync()
+
+      val result = cloudService.verifyAndSend(10L, "otherHash", IO.delay(List(().asRight[Throwable]))).unsafeRunSync()
+      val expected = List(Left(PrecedingSnapshotIsNotThePreviousOne(previousSnapshot, 10L, "otherHash")))
+
+      result shouldBe expected
+    }
+  }
+
+  "enqueueSnapshot" - {
+    "should enqueue, process snapshot and the last sent snapshot should be set correctly" in {
+      cloudServiceEnqueue.enqueueSnapshot(File(""), 10L, "somehash").unsafeRunSync()
+      cloudServiceEnqueue.enqueueSnapshotInfo(File(""), 10L, "somehash").unsafeRunSync()
+      cloudServiceEnqueue.enqueueSnapshotInfo(File(""), 12L, "otherHash").unsafeRunSync()
+      Thread.sleep(50) //to let the stream finish
```

Comment:
Those two thread sleeps will increase the duration of the build by 100 seconds. There should be a better way of doing this, I'm sure that ScalaTest offers some polling functionality, so that you can actively await for a condition to be satisfied.